### ISSUE TITLE
exists.elim in tactics mode

### ISSUE THREAD

**Issue by htzh** [leanprover/tutorial#100]
I can use ```obtain _ _, from _``` outside of tactics mode, but I can't figure out how to use it inside. If I don't use tactics I don't know how to force expansion of definitions. Also I have no feedback on what the current context and goal look like. Once inside tactics it seems really hard to introduce new intermediate facts into the context. The particular case I am stuck on is
```
import algebra.group data.set data.fintype
namespace algebra
open set
namespace set
open function
variables {A B C: Type}
variable {S : set A}
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g,
      setext (assume z,
      iff.intro
        (begin
        rewrite [↑compose, ↑image, ↑set_of],
        intro Hz,
        --rewrite 
        end) _)
end set
```

**Comment by htzh** [comment]
I made some progress, but it still doesn't type check at ```show```. I manage to get some feedback by inserting spurious ```begin end``` pair. But once I am inside ```exists.intro``` I can no longer tell what is expected.
```
namespace set
open function
variables {A B C: Type}
variable {S : set A}
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g,
      setext (assume z,
      iff.intro
        (begin
        rewrite [↑compose, ↑image, ↑set_of, ↑mem],
        intro Hz,
        show ∃ (x : A), S x ∧ f (g x) = z, from
        (obtain y Hy, from Hz, obtain x Hx, from and.left Hy,
        exists.intro x (and.intro (and.left Hx) (eq.subst (eq.symm (and.right Hx)) (and.right Hy))),        
        begin end
        )
        --rewrite 
        end) _)
end set
```

**Comment by htzh** [comment]
By removing the spurious ```begin end``` it seemed able to type check at ```show```. Why is it illegal to have an empty pair? It is rather painful to try to get good feedback -- seems like it would be nice to have the emacs mode behave like as if there is always a ```begin end``` pair at the point (or something like that) to show the current context and goal.

**Comment by htzh** [comment]
It finally type checked but it certainly would have helped to get more feedback in the process:
```
import algebra.group data.set data.fintype
namespace algebra
open set
namespace set
open function
variables {A B C: Type}
variable {S : set A}
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g,
      setext (assume z,
      iff.intro
        (begin
        rewrite [↑compose, ↑image, ↑set_of, ↑mem],
        intro Hz,
        show ∃ (x : A), S x ∧ f (g x) = z, from
        (obtain y Hy, from Hz, obtain x Hx, from and.left Hy,
        exists.intro x (and.intro (and.left Hx) (eq.subst (eq.symm (and.right Hx)) (and.right Hy)))        
        )
        end)
        (begin
        rewrite [↑compose, ↑image, ↑set_of, ↑mem],
        intro Hex,
        show ∃ (x : B), (∃ (x_1 : A), S x_1 ∧ g x_1 = x) ∧ f x = z, from (
        obtain x Hx, from Hex,
        exists.intro (g x) (and.intro (exists.intro x (and.intro (and.left Hx) rfl)) (and.right Hx))
             )
        end))
end set
```

**Comment by htzh** [comment]
I cleaned up with better indentation and more type annotation. Maybe this can be added to data.set.function in the library. There seems to be another track of developing theories of image using subtype.
```
import data.set
namespace algebra
open set
namespace set
open function
section
variables {A B C: Type}
variable {S : set A}
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g,
      setext (assume z,
      iff.intro
        (begin
        rewrite [↑compose, ↑image, ↑set_of, ↑mem],
        intro Hz,
        show ∃ (x : A), S x ∧ f (g x) = z, from
        (obtain y (Hy : (∃ (x : A), S x ∧ g x = y) ∧ f y = z), from Hz,
         obtain x (Hx : S x ∧ g x = y), from and.left Hy,
         exists.intro x
           (and.intro (and.left Hx)
                      (eq.subst (eq.symm (and.right Hx)) (and.right Hy))))
        end)
        (begin
        rewrite [↑compose, ↑image, ↑set_of, ↑mem],
        intro Hex,
        show ∃ (x : B), (∃ (x_1 : A), S x_1 ∧ g x_1 = x) ∧ f x = z, from
        (obtain x (Hx : S x ∧ f (g x) = z), from Hex,
         exists.intro (g x)
         (and.intro (exists.intro x
                                   (and.intro (and.left Hx) rfl))
                     (and.right Hx)))
        end))
end
end set
```
Once it type checks I seem to get more type information clicking on symbols but it is still not consistent. For example the Hx in the first iff branch does not get any type information in Emacs.

**Comment by fpvandoorn** [comment]
Do you know that if you write a term with underscores, then you can see the expected type of those underscores by hovering over them (in Emacs) or compiling your file? So if I write a proof in declarative style (no tactics), I do so incrementally. For example, first
```lean
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g, _
```
then
```lean
lemma image_compose : ∀ (f : B → C) (g : A → B), f '[g '[S]] = (f ∘ g) '[S] :=
      assume f, assume g,
      setext (assume z,
      iff.intro _ _)
```
and so on.

This gives very similar feedback as tactic proofs: hovering over the underscores gives the current "goal".

To your other questions: I don't know whether there is a way to use the `obtain` command in tactic mode. However, you can use cases, which does the same. Example:
```lean
open nat
variables (P Q : ℕ → Prop)

example (H1 : ∃n, P n) (H2 : ∀n, P n → Q n) : ∃n, Q n :=
begin
  cases H1 with n p,
  apply (exists.intro n),
  apply H2,
  exact p
end
```

Also, in tactic mode you can use `assert` to add a new fact to the context. Example:
```lean
open nat
variables (P Q : ℕ → Prop)

example (H1 : ∃n, P n) (H2 : ∀n, P n → Q n) : ∃n, P n ∧ Q n :=
begin
  cases H1 with n p,
  apply (exists.intro n),
  assert q : Q n,
    {apply H2, exact p},
  apply and.intro, exact p, exact q
end
```

> By removing the spurious begin end it seemed able to type check at show. Why is it illegal to have an empty pair?

A `begin end` pair is interpreted as a term. If you don't write any tactic in a `begin end` pair, it is a metavariable. So `f a b begin end` is interpreted as `f a b _`, which is not type-correct if `f` takes only 2 arguments.

> I cleaned up with better indentation and more type annotation. Maybe this can be added to data.set.function in the library.

@avigad: can you do this/respond to this?

**Comment by avigad** [comment]
Dear Haitao,

Thanks for the proof! I will add it to the library. I revised it a little to make it more readable:
```lean
import data.set
open set function eq.ops

variables {A B C : Type} {S : set A}

theorem in_image {x : A} {y : B} {f : A → B} 
  (H1 : x ∈ S) (H2 : f x = y) : y ∈ f '[S] :=
exists.intro x (and.intro H1 H2)

lemma image_compose (f : B → C) (g : A → B) : f '[g '[S]] = (f ∘ g) '[S] :=
setext (take z,
  iff.intro
    (assume Hz : z ∈ f '[g '[S]],
      obtain y (Hy : y ∈ g '[S] ∧ f y = z), from Hz,
      obtain x (Hz : x ∈ S ∧ g x = y), from and.left Hy,
      show z ∈ (f ∘ g) '[S], from in_image (and.left Hz) ((and.right Hz)⁻¹ ▸ (and.right Hy)))
    (assume Hz : z ∈ (f ∘ g) '[S],
      obtain x (Hx : x ∈ S ∧ f (g x) = z), from Hz,
      have Hgx : g x ∈ g '[S], from in_image (and.left Hx) rfl,
      show z ∈ f '[g '[S]], from in_image Hgx (and.right Hx)))
```

Some notes: 

- if you say ```namespace set``` then ```open set``` is not needed.
- an equivalent form of ```rewrite [↑compose, ↑image, ↑set_of, ↑mem]``` is ```esimp [compose, image, set_of, mem]```
- in fact, these are not needed in the proof above (the unifier unfolds the definitions)

**Comment by htzh** [comment]
@fpvandoorn Thanks for the very detailed explanations. I have used the underscore trick before but it seems most of the time I get the "can't synthesize placeholder" error. It is possible I had other type errors that prevented the feedback. I will give it more practice. I like to use the declarative style if I can. Thanks for showing me ```cases```. I also discovered ```match ... with ...``` which is more verbose but clearer? But ```cases``` looks good for unambiguous situations. The tutorial doesn't make it clear that ```assert``` is also a tactic itself. I have been creating ```show _, from _``` just so that I can use ```assert```.

@avigad Thanks for the refactoring and showing me a more idiomatic presentation.
